programs/bpf_loader/src/syscalls.rs

```diff
@@ -1329,6 +1343,65 @@ impl<'a> SyscallObject<BpfError> for SyscallMemset<'a> {
     }
 }
 
+// Blake3
+pub struct SyscallBlake3<'a> {
+    base_cost: u64,
+    byte_cost: u64,
+    compute_meter: Rc<RefCell<dyn ComputeMeter>>,
+    loader_id: &'a Pubkey,
+}
+impl<'a> SyscallObject<BpfError> for SyscallBlake3<'a> {
+    fn call(
+        &mut self,
+        vals_addr: u64,
+        vals_len: u64,
+        result_addr: u64,
+        _arg4: u64,
+        _arg5: u64,
+        memory_mapping: &MemoryMapping,
+        result: &mut Result<u64, EbpfError<BpfError>>,
+    ) {
+        question_mark!(self.compute_meter.consume(self.base_cost), result);
+        let hash_result = question_mark!(
```

Comment:
Should we ensure that the size of `hash_result` matches `blake3::HASH_BYTES` ?

Reply:
@t-nelson I am not sure what I should do here. There are 12 uses of translate_slice_mut in this file and none of them (including the sha256 and keccak syscalls) check that the result is the length that you pass as an argument. Based on this observation (unless I'm incorrect) I would say that I do not need to ensure that the has_result matches blake3::HASH_BYTES.

Reply:
Because the result is expected to be fixed width, the syscall is assuming that the buffer passed in is at least the size of the result that the syscall will write out.  This is checked by the translate call.  The translate call does not check anything other than the fact that the program has rw access to as many byes as the syscall intends to write.

Reply:
This doesn't specifically protect the program from being dumb, but does protect the runtime from being compromised.

Reply:
Got it. So worst case the calling program panics, not the validator process.  Thanks for the explainer :slightly_smiling_face: